src/Workspaces/Core/Portable/Remote/RemoteArguments.cs

```diff
@@ -37,35 +37,31 @@ public FindReferencesSearchOptions Rehydrate()
         }
     }
 
-    internal class SerializableSymbolAndProjectId : IEquatable<SerializableSymbolAndProjectId>
+    internal class SerializableSymbolAndProjectId
     {
         public string SymbolKeyData;
         public ProjectId ProjectId;
 
-        public override int GetHashCode()
-            => Hash.Combine(SymbolKeyData, ProjectId.GetHashCode());
-
-        public override bool Equals(object obj)
-            => Equals(obj as SerializableSymbolAndProjectId);
-
-        public bool Equals(SerializableSymbolAndProjectId other)
-            => other != null && SymbolKeyData.Equals(other.SymbolKeyData) && ProjectId.Equals(other.ProjectId);
-
         public static SerializableSymbolAndProjectId Dehydrate(
-            IAliasSymbol alias, Document document)
+            IAliasSymbol alias, Document document, CancellationToken cancellationToken)
         {
             return alias == null
                 ? null
-                : Dehydrate(new SymbolAndProjectId(alias, document.Project.Id));
+                : Dehydrate(document.Project.Solution, alias, cancellationToken);
         }
 
         public static SerializableSymbolAndProjectId Dehydrate(
-            SymbolAndProjectId symbolAndProjectId)
+            Solution solution, ISymbol symbol, CancellationToken cancellationToken)
         {
+            var symbolKey = symbol.GetSymbolKey(cancellationToken);
+            var projectId = solution.GetExactProjectId(symbol);
+            if (projectId == null)
+                throw new ArgumentException("Symbol's project could not be found in the Solution provided");
```

Comment:
>throw new ArgumentException [](start = 16, length = 27)

Use `Contract.NotNull(projectId, "Symbol's project could not be found in the Solution provided");` since this is not public API.

Reply:
so, general concern: i want to do this check for all public entrypoints.  having to remember to do it at every entrypoint seems unpleasant. This centralizes.  Is there a good solution for that?

Reply:
What are those public entry points?

---
In reply to: [411610773](https://github.com/dotnet/roslyn/pull/43207#discussion_r411610773) [](ancestors = 411610773)

Reply:
All the public SymbolFinder methods.  Like FindReferences, etc.

Reply:
note: in the "remove symbolandprojectid" pr this becomes a lot simpler (because we cna merge tons of these methods and get rid of the internal ones).  In that PR i can add this check to our public facing entrypoints, and make this a contract method inside just in case.

Reply:
@tmat i have added the public-entrypoint checking in the https://github.com/dotnet/roslyn/pull/43483 PR.  Thanks!

Reply:
Good. Our internal code does not expect arbitrary exceptions to fly around. It may report Watsons for them, which we definitely don't want for the case when public API input is incorrect. Another problem of ArgumentExceptions in the middle of internal implementation is missing context at the throw site. We don't know what argument was passed thru public API that caused the error condition. Was there an argument at all or was it perhaps just invalid state of some object. We might say it's obvious now but as code evolves nothing is keeping the relationship of the throw site to the public entry point. Then the user might end up with exceptions that do not make any sense.

Reply:
yeah, i agree :)   next PR will have the right code here (i.e. checking and throwing at all the actual public entrypoints) :)